### PR TITLE
[slackbot] strip away citation marker during streaming 

### PR DESCRIPTION
**Pull request by ykmsd**
## Description
This PR is to avoid showing citation maker while slackbot is streaming. If streaming token contains a citation marker, we skip it. If it ends with something look like a citation marker (e.g. `:cite[`, we keep it, and when the next token arrives: 
  - If it completes a citation => the full marker gets stripped (not rendered)
  - If it doesn't =>  the held-back text gets prepended and rendered normally

Once streaming is done, we format the markdown and process citation here so we should be able to show the proper citation: https://github.com/dust-tt/dust/blob/a20015445b0c606c64ce499633b953e4361345c9/connectors/src/connectors/slack/chat/stream_conversation_handler.ts#L467-L482

Closing https://github.com/dust-tt/tasks/issues/7756

## Tests

Tested locally with @matteotrab, it seems to be working but we might miss some edge cases. 

## Deploy Plan

---

<a href="https://stagereview.app/dust-tt/dust/pull/25497">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 1 individual chapter for you:

| | Title |
|---|-------|
| 1 | [Strip citation markers during Slack streaming](https://stagereview.app/dust-tt/dust/pull/25497/chapters/1) |

<a href="https://stagereview.app/dust-tt/dust/pull/25497">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 31817c90a8515cdd8976e1d02209d44751885fbf on May 12, 2026 10:07am UTC.</sub>

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts [hunk 1]
```diff
@@ -433,7 +435,24 @@ async function streamAgentAnswerToSlack(
...
+          // Strip complete :cite[...] markers (resolved to footnotes on final chat.update).
+          let safeText = combined.replace(/ ?:cite\[[a-zA-Z0-9, ]+\]/g, "");
+
+          // Hold back trailing partial markers until the next token.
+          const partialMatch = safeText.match(
+            / ?:c(?:i(?:t(?:e(?:\[[a-zA-Z0-9, ]*)?)?)?)?$/
+          );
+          if (partialMatch) {
+            pendingCitePrefix = partialMatch[0];
+            safeText = safeText.slice(0, -pendingCitePrefix.length);
```
Comment (ykmsd) [inline connectors/src/connectors/slack/chat/stream_conversation_handler.ts:450]: if it's a partial match, it's always the end of the token, so we slice from 0 to the beginning of partial citation.

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts [hunk 2]
```diff
@@ -433,7 +435,24 @@ async function streamAgentAnswerToSlack(
...
-          await streamHandler.appendText(event.text);
+          const combined = pendingCitePrefix + event.text;
+          pendingCitePrefix = "";
+
+          // Strip complete :cite[...] markers (resolved to footnotes on final chat.update).
+          let safeText = combined.replace(/ ?:cite\[[a-zA-Z0-9, ]+\]/g, "");
+
+          // Hold back trailing partial markers until the next token.
+          const partialMatch = safeText.match(
+            / ?:c(?:i(?:t(?:e(?:\[[a-zA-Z0-9, ]*)?)?)?)?$/
```
Comment (davidebbo) [inline connectors/src/connectors/slack/chat/stream_conversation_handler.ts:446]: That's kind of a scary regex. Maybe better handled in a different way?

**Review by spolu — APPROVED** [review]
LGTM

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts (hunk 1 above)
Comment (davidebbo) [reply connectors/src/connectors/slack/chat/stream_conversation_handler.ts:450]: What prevents it from matching any word that starts with `c` (for instance)? Also, how would we handle the second half of the citation in the next chunk?

I may be completely misunderstanding this...

Comment (davidebbo) [reply connectors/src/connectors/slack/chat/stream_conversation_handler.ts:450]: Ok, discussed IRL

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts [hunk 3]
```diff
@@ -433,7 +435,24 @@ async function streamAgentAnswerToSlack(
           break;
         }
         if (answer.length <= MAX_SLACK_MESSAGE_LENGTH) {
-          await streamHandler.appendText(event.text);
+          const combined = pendingCitePrefix + event.text;
+          pendingCitePrefix = "";
+
+          // Strip complete :cite[...] markers (resolved to footnotes on final chat.update).
+          let safeText = combined.replace(/ ?:cite\[[a-zA-Z0-9, ]+\]/g, "");
```
Comment (davidebbo) [inline connectors/src/connectors/slack/chat/stream_conversation_handler.ts:442]: As discussed, maybe we don't want to check for space at all in the beginning (for both regexes)

**Review by davidebbo — APPROVED** [review]
LGTM with comments

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts (hunk 3 above)
Comment (ykmsd) [reply connectors/src/connectors/slack/chat/stream_conversation_handler.ts:442]: checked IRL, we need to keep optional spacing, because if a citation maker appears in the middle of the sentence, we need to remove space together, otherwise we will have double space. But it should be optional in case a citation marker appears in the beginning of the sentence.
